lib/cn/kcompact.c

```diff
@@ -175,176 +79,214 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
+                    w->cw_stats.ms_val_bytes_out += vlen;
+                    w->cw_vbmap.vbm_used += vlen;
+                }
+            } else {
+                /* The only time we ever land here is when the same
+                 * key appears in two input kvsets with overlapping
+                 * sequence numbers.  For example:
+                 *
+                 * Kvset #1: MEAL --> [[11,BURGERS], [10,BEER]]
+                 * Kvset #2: MEAL --> [[10,SPAM], [9,FRIES]]
+                 *
+                 * We have already emitted BURGERS and BEER, are
+                 * currently processing SPAM (which must be be
+                 * discarded), and will get FRIES on the next
+                 * iteration.
+                 *
+                 * The following assertions verify that we aren't here
+                 * for other reasons (e.g., input kvsets that violate
+                 * assumptions about sequence numbers).
+                 */
+                assert(dbg_prev_idx < idx);
+                assert(seq == emitted_seq);
+                if (seq > emitted_seq)
+                    seqno_errcnt++;
+
+                assert(vdata != HSE_CORE_TOMB_PFX);
             }
-            if (ev(err))
-                goto done;
-            emitted_val = true;
-
-            if (vtype == vtype_ptomb)
-                emitted_seq_pt = seq;
-            else
-                emitted_seq = seq;
-
-            w->cw_stats.ms_val_bytes_out += complen ? complen : vlen;
-            w->cw_vbmap.vbm_used += complen ? complen : vlen;
-        } else {
-            /* The only time we ever land here is when the same
-             * key appears in two input kvsets with overlapping
-             * sequence numbers.  For example:
-             *
-             * Kvset #1: MEAL --> [[11,BURGERS], [10,BEER]]
-             * Kvset #2: MEAL --> [[10,SPAM], [9,FRIES]]
-             *
-             * We have already emitted BURGERS and BEER, are
-             * currently processing SPAM (which must be be
-             * discarded), and will get FRIES on the next
-             * iteration.
-             *
-             * The following assertions verify that we aren't here
-             * for other reasons (e.g., input kvsets that violate
-             * assumptions about sequence numbers).
-             */
-            assert(dbg_prev_src < curr.src);
-            assert(seq == emitted_seq);
-            if (seq > emitted_seq)
-                seqno_errcnt++;
-
-            assert(vdata != HSE_CORE_TOMB_PFX);
         }
-    }
 
-    prev_kobj = curr.kobj;
+        prev_kobj = curr->kobj;
 
-    dbg_dup = false;
-    dbg_nvals_this_key = 0;
-    dbg_prev_src = curr.src;
+        dbg_dup = false;
+        dbg_nvals_this_key = 0;
+        dbg_prev_idx = idx;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (ev(err))
-        goto done;
+        /* Discard the result from pop(). This bin heap is backed by a kv
```

Comment:
Gaurav let me know if you have a better way to word this.

Also you suggested maybe creating a function for this usage of pop(). Did you have any ideas for a name?

Reply:
bin_heap2_discard() ?  Then it just wouldn't take an item pointer so no need to copy out the top item...

Reply:
I ended up just letting pop() take a NULL item argument. To implement discard(), you would probably do it by calling pop() anyways, so 🤷